### ISSUE TITLE
[Cpp] target  fails to recognize rule alternatives in certain nested structures

### ISSUE THREAD

**Issue by mlangiu** [antlr/antlr4#1753]
First of all: I really appreciate the great work and all the effort that have been put into ANTLR, both the original @parrt and the Cpp target of @mike-lischke. I have been able to learn a lot from both and managed to get pretty far with an Expression parser I am building with the Cpp target.
However, I encountered a general issue that makes continuation with my project impossible (at least in the way I intended it).
Unfortunately, I am not able to patch it myself as I frankly don't comprehend most of the internals  that are happening during rule alternative matching.    

**Expected behavior:**
Cpp target yields identical parse results as Java target

**Actual behavior:**
In grammars with certain nested structures, the Cpp target only recognizes the first of a number of similar alternatives.

**MWE grammar:**
While the Java target parses all 'booleanExpression' alternatives in the below grammar, the Cpp target will only accept the first one listed.
If the other cases occur, it matches a 'numericalExpression' and ignores all following tokens (despite not trowing any token recognition errors!)

This behavior can be prevented by certain modifications to the grammar which are noted in the comments contained therein, however, these modifications are obviously limiting the grammar. 

```
grammar Exptree;

///////////
// Rules //
///////////

// If the 'numericalExpression' alternative is removed, all booleanExpressions 
// are recognized
start
: booleanExpression
| numericalExpression
;

// The order here seems to have no effect on what is being recognized
booleanExpression
: greater
| greaterEqual
| less
| lessEqual
| identity
| distinction
;

// For some reason only the FIRST alternative listed here will be recognized! 
// All others are falsely interpreted as numericalExpressions
distinction
: numericalExpression (NE  numericalExpression)+
;
identity
: numericalExpression (EQ  numericalExpression)+
;
greaterEqual
: numericalExpression (GE  numericalExpression)+
;
lessEqual
: numericalExpression (LE  numericalExpression)+
;
greater
: numericalExpression (G  numericalExpression)+
;
less
: numericalExpression (L  numericalExpression)+
;

// if this rule is removed and all occurrences of 'numericalExpression' 
// are replaced with 'NUMBER' all alternatives above are recognized!
numericalExpression
: NUMBER
;

////////////
// Tokens //
////////////

NUMBER
: DIGIT+
;

G
: '>'
;

GE
: '>='
;

L
: '<'
;

LE
:'<='
;

EQ
: '=='
;

NE
: '!='
;

DIGIT
: [0-9]
;
   
WS
: [ \t\r\n]+ -> skip
;

```
Generation of the Cpp target code was done (On a Mac running OS 10.12) with 
`
LOCATION=/usr/local/lib/antlr-4.6-complete.jar
java -jar $LOCATION -Dlanguage=Cpp -listener -visitor -o cpp/ -package Exptree Exptree.g4
`
The main.cpp I used, looks as shown below, the build was done using Xcode 8.2.1.   

```
#include "ExptreeLexer.h"
#include "ExptreeParser.h"
#include "ExptreeBaseVisitor.h"
#include "ExptreeBaseListener.h"

#include <sstream>

using namespace Exptree;
using namespace antlr4;
using std::string;
using std::cout;
using std::endl;

int main(int argc, const char * argv[]) {
    
    if (argc == 2) {
        const char *filename;
        
        filename = argv[1];
        std::ifstream strm( filename );
        
        if (strm.is_open()) {
            ANTLRInputStream input(strm);
            ExptreeLexer lexer(&input);
            CommonTokenStream tokens(&lexer);
            
            tokens.fill();
            cout << "INPUT:" << endl;
            for (auto token : tokens.getTokens()) {
                //DEBUG print out recognized tokens
                cout << token->toString() << endl;
            }
            
            ExptreeParser parser(&tokens);
            tree::ParseTree *tree = parser.start();
            
            cout << "\n";
            
            cout << "ANTLR INTERPRETATION:" << endl;
            cout << tree->toStringTree(&parser) << endl;
            
            cout << "\n";
            return 0;
        }
    }
    
    return 0;
}
```

**Comment by mike-lischke** [comment]
Can you also add the example input you used for testing?

**Comment by mlangiu** [comment]
Certainly, the input is simply a text file called input.txt residing in the same directory as the grammar.
Contents that are parsed correctly are in the above case:
`
1 != 0
`
Contents that are not parsed correctly (but are exprected to) are any of the following:
`
1 == 0
`
`
1 >= 0
`
`
1 <= 0
`
`
1 > 0
`
`
1 < 0
`

**Comment by pureconfig** [comment]
@mlangiu that grammar is bogus, and fails even in IntelliJ's previewer. You parse the entire file, containing several lines of input, yet you -> skip whitespaces.

**Comment by mlangiu** [comment]
@pureconfig I'm afraid I don't know what you mean.  
The Java build I am referring to was done with the command line and produces the desired results:
![workingparse1](https://cloud.githubusercontent.com/assets/15340151/23857150/77f9d954-07fc-11e7-9f69-ecc4d63b9ea1.png)
![workingparse2](https://cloud.githubusercontent.com/assets/15340151/23857149/77f6804c-07fc-11e7-93ce-20427e1bd0f6.png)

**Comment by pureconfig** [comment]
Then the Java build works by accident. Fix the grammar.

**Comment by mlangiu** [comment]
Could you be a bit more specific? What did you mean by 'multiple lines'
Maybe you are confused by the extra newlines between alternatives, I have found the notation
`
rule : a | b;
`
to be equivalent to the notation
`
rule
: a
| b
;
` 
As I mentioned, the input of the Cpp version is a file containing a single line. The file location is passed to the main call as a command line argument and processed as evident from the main.cpp.
The Java test was done with the basic command line interface. There I provided the input manually over stdin.
In any case I don't see how the way I am handing over the input has to do anything with the grammar.

**Comment by parrt** [comment]
Grammar is fine but only matches a single expression; is that the confusion @pureconfig ?

**Comment by pureconfig** [comment]
@parrt Yeah, his test program parses a file with multiple lines.

**Comment by pureconfig** [comment]
@mlangiu paste your input.txt into IntelliJ's preview. Does it look right to you?

**Comment by pureconfig** [comment]
@mlangiu OK, I assumed the input.txt contained the lines you posted above.

**Comment by pureconfig** [comment]
@mlangiu I generated a parser based on your grammar, and all single-line inputs works as expected.

**Comment by mlangiu** [comment]
@pureconfig I'm assuming you mean that this grammar works fine with the Java target? 
That's what I already noted above, the issue is that the C++ tagret it doesn't! 
You can refer to my first post, more specifically the comments within the grammar to see what is going wrong.

**Comment by pureconfig** [comment]
@mlangiu are you saying the output from the Cpp-target I posted above is wrong?

**Comment by mlangiu** [comment]
What output are you referring to?

**Comment by pureconfig** [comment]
Uh, when I run the test program you posted against the grammar you posted, on the Cpp target, I get:

```
INPUT:
[@0,0:0='1',<1>,1:0]
[@1,2:2='>',<2>,1:2]
[@2,4:4='0',<1>,1:4]
[@3,6:5='<EOF>',<-1>,2:0]

ANTLR INTERPRETATION:
(start (booleanExpression (greater (numericalExpression 1) > (numericalExpression 0))) <EOF>)
```

```
INPUT:
[@0,0:0='1',<1>,1:0]
[@1,2:3='!=',<7>,1:2]
[@2,5:5='0',<1>,1:5]
[@3,7:6='<EOF>',<-1>,2:0]

ANTLR INTERPRETATION:
(start (booleanExpression (distinction (numericalExpression 1) != (numericalExpression 0))) <EOF>)
```

```
[@0,0:0='1',<1>,1:0]
[@1,2:2='<',<4>,1:2]
[@2,4:4='0',<1>,1:4]
[@3,6:5='<EOF>',<-1>,2:0]

ANTLR INTERPRETATION:
(start (booleanExpression (less (numericalExpression 1) < (numericalExpression 0))) <EOF>)
```

```
INPUT:
[@0,0:0='1',<1>,1:0]
[@1,2:3='==',<6>,1:2]
[@2,5:5='0',<1>,1:5]
[@3,7:6='<EOF>',<-1>,2:0]

ANTLR INTERPRETATION:
(start (booleanExpression (identity (numericalExpression 1) == (numericalExpression 0))) <EOF>)
```

Which looks correct.

**Comment by mlangiu** [comment]
Sorry, I didn't see you edited your post above.
I am using the Xcode project from the repo to build the library for the ANTLR runtime (libantlr4-runtime.a). 
Then I include it into another Xcode project to build the Expressiontree and custom source files with.

My results yield 
For input `1 != 0` (The first one explicitly specified `booleanExpression` in the grammar):
```
INPUT:
[@0,0:0='1',<1>,1:0]
[@1,2:3='!=',<7>,1:2]
[@2,5:5='3',<1>,1:5]
[@3,7:6='<EOF>',<-1>,2:0]

ANTLR INTERPRETATION:
(start (booleanExpression (distinction (numericalExpression 1) != (numericalExpression 3))))
```
For input `1 == 0` (and all others `booleanExpression`s equivalently):
```
INPUT:
[@0,0:0='1',<1>,1:0]
[@1,2:3='==',<6>,1:2]
[@2,5:5='3',<1>,1:5]
[@3,7:6='<EOF>',<-1>,2:0]

ANTLR INTERPRETATION:
(start (numericalExpression 1))
```

Your results are indeed what is intended, but the `<EOF>` token at the end of your output surprises me! 
Did you modify anything? What did you use to build your runtime / get the posted output?

**Comment by pureconfig** [comment]
@mlangiu No, I did not modify anything, I used what you posted verbatim. 

I generated the parser with the latest plugin (used inside CLion), and built with clang from CLion.

**Comment by mlangiu** [comment]
I did indeed use the HEAD as far as I remember, I will check whether my issue is due to this or if it is an Xcode problem and report back tomorrow when I get back to work.
Thanks for the hint!

**Comment by pureconfig** [comment]
@mike-lischke maybe worth checking out if there's a regression on master? The released binaries definitely works just fine.

**Comment by mike-lischke** [comment]
Very nice seeing you analyzing this already. I'll check the latest sources, but our unit tests didn't show anything wrong. Strange...

**Comment by mlangiu** [comment]
Alright, so the issue appears to be caused by the use of the Xcode Project in 
https://github.com/antlr/antlr4/tree/master/runtime/Cpp/runtime
I checked out the version I am using on January 22. and when using it within my own project to build the static ANTLR library, the behavior described above occurs.

Using the runtime provided on the website (http://www.antlr.org/download/antlr4-cpp-runtime-4.6-macos.zip) I managed to get correct looking results both in the example grammar and my actual project.

The reason I checked out from head was because the first mention of the C++ target on the download page links to the Documentation for the C++ target in this repository.
Under 'Where can I get the runtime' it says:

> Instead of downloading a prebuilt binary you can also easily build your own library on OSX or Windows. Just use the provided projects for XCode or Visual Studio and build it.

Despite the mention of the _prebuilt runtime_ I built my own, because the link provided simply links back to the landing page of the ANTLR Website, so I naively assumed that this was some legacy from times where the runtimes were directly listed there.

To sum up: 
- When using the provided C++ runtime at the bottom of the download page of the ANTLR website, things work as expected.
- To make the doc absolutely foolproof I suggest making the link to the prebuilt binaries point directly to the download section or even better list the runtime downloads directly under each respective documentation. 
- The runtime built with the repo head _still causes the abovementioned issues_ for me. If there are more things you need to know from me in order to fix this let me know ;)  

Thanks for the help

**Comment by parrt** [comment]
@mike-lischke can you confirm the "*The runtime built with the repo head still causes the abovementioned issues for me.*"

**Comment by pureconfig** [comment]
@parrt I'm happy to test the current HEAD as well - I use the Cpp runtime for a very complex grammar. I'll test @mlangiu 's grammar as well.

**Comment by pureconfig** [comment]
It seems like the new runtime requires me to implement getChannelNames. What's that about?

**Comment by pureconfig** [comment]
What's the new getChannelNames() method about?

**Comment by parrt** [comment]
nah. it's for HIDDEN channels etc...

**Comment by pureconfig** [comment]
@parrt Are the 4.5, 4.6, etc branches *exactly* what makes it to the corresponding releases?

**Comment by parrt** [comment]
I believe so maybe minus a doc file tweak or something.

**Comment by mike-lischke** [comment]
I'm afraid I cannot keep up with you. I'll try to take a closer look next weekend. Almost no time during the week.

**Comment by pureconfig** [comment]
@mlangiu I would be great if you could try again with the *current* HEAD.

**Comment by mlangiu** [comment]
Just did, but I am getting the error that my generated Lexer is abstract:

> main.cpp:24:26: Variable type 'Exptree::ExptreeLexer' is an abstract class

I am using the grammar and main from my first post... Is there some new way to get a CommonTokenStream that doesn't rely on instantiating a Lexer?

**Comment by mike-lischke** [comment]
OK, I just tried with latest ANTLR code from github and everything looks fine, just like what @pureformat found. IMO everything is as it should be. @mlangiu try also the new 4.7 release and see if your grammar problems are gone now. If that's the case we can close this issue.

**Comment by mlangiu** [comment]
Sorry it took me a while to get to try it, but yes, the current release works as expected!
Thanks eeryone for the great work! :)

**Comment by mike-lischke** [comment]
Great, thanks for the confirmation.

**Comment by mike-lischke** [comment]
@mlangiu Please close this bug report if the issue is solved for you.
